src/arch/x86/x86_64/src/bios.rs

```diff
@@ -0,0 +1,380 @@
+    pub reserved: u16,      /* reserved - must be zero */
+    pub local_apic_id: u32, /* Processor x2APIC ID  */
+    pub lapic_flags: u32,
+    pub uid: u32, /* ACPI processor UID */
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtInterruptOverride {
+    pub header: AcpiSubtableHader,
+    pub bus: u8,
+    pub sourceirq: u8,
+    pub globalirq: u32,
+    pub flags: u16,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtNMI {
+    pub header: AcpiSubtableHader,
+    pub flags: u16,
+    pub globalirq: u32,
+}
+
+#[repr(packed)]
+#[derive(Default)]
+pub struct AcpiMadtLAPICNMI {
+    pub header: AcpiSubtableHader,
+    pub acpi_processor_uid: u8,
+    pub flags: u16,
+    pub local_interrupt: u8,
+}
+
+fn write<T>(w: &mut print::WriteTo, val: T, offset: usize, index: usize) {
+    let y = (offset + index * size_of::<T>()) as *mut T;
+    unsafe {
+        ptr::write_volatile(y, val);
+    }
+    write!(w, "write to {:x?}: \r\n", y).unwrap();
+}
+
+fn read<T>(offset: usize, index: usize) -> T {
+    let y = (offset + index * size_of::<T>()) as *mut T;
+    let v = unsafe { ptr::read_volatile(y) };
+    v
+}
+
+/*
+ *
+ * Intel ACPI Component Architecture
+ * ASL Optimizing Compiler version 20140214-64 [Mar 29 2014]
+ * Copyright (c) 2000 - 2014 Intel Corporation
+ *
+ * Compilation of "vmm_acpi_dsdt.dsl" - Fri Apr  1 13:34:26 2016
+ *
+ */
+
+/*
+ *
+ * Based on the example at osdev wiki wiki.osdev.org/AML,
+ * and the other example in http://www.acpi.info/DOWNLOADS/ACPI_5_Errata%20A.pdf
+ * on page 194
+ *
+ * Compiled with `iasl -sc input_file.dsl`
+ */
+
+/*
+ *       9:  DefinitionBlock (
+ *      10:      "vmm_acpi_dsdt.aml", // Output AML Filename : String
+ *      11:      "DSDT",              // Signature : String
+ *      12:      0x2,                 // DSDT Compliance Revision : ByteConst
+ *      13:      "MIKE",              // OEMID : String
+ *      14:      "DSDTTBL",           // TABLE ID : String
+ *      15:      0x0                  // OEM Revision : DWordConst
+ *      16:  ){}
+ */
+
+pub const DSDT_DSDTTBL_HEADER: [u8; 36] = [
+    0x44, 0x53, 0x44, 0x54, 0x24, 0x00, 0x00, 0x00, /* 00000000    "DSDT$..." */
+    0x02, 0xF3, 0x4D, 0x49, 0x4B, 0x45, 0x00, 0x00, /* 00000008    "..MIKE.." */
+    0x44, 0x53, 0x44, 0x54, 0x54, 0x42, 0x4C, 0x00, /* 00000010    "DSDTTBL." */
+    0x00, 0x00, 0x00, 0x00, 0x49, 0x4E, 0x54, 0x4C, /* 00000018    "....INTL" */
+    0x14, 0x02, 0x14, 0x20, /* 00000020    "... " */
+];
+
+#[inline]
+fn gencsum(start: usize, end: usize) -> u8 {
+    let mut tot: u16 = 0;
+    for i in start..end + 1 {
+        let b: u8 = read(i, 0);
+        tot = tot + b as u16;
+    }
+    tot as u8
+}
+
+// I still don't know why these were different, I thought it was the
+// same algorithm.
+#[inline]
+fn acpi_tb_checksum(start: usize, end: usize) -> u8 {
+    // I don't understand this one
+    // data.iter().map(|x| Wrapping(*x)).sum::<Wrapping<u8>>().0
```

Comment:
That is functional programming style, and though and don't know Rust that well yet nor the data structures here, this is what I read - in very short:
- get a different view on `data` by calling `iter()` on it; looks like that converts the underlying data structure to something that can be operated on
- apply a `map` - for every element in the previous set, create a new one; i.e., the number of elements stays the same. `Wrapping(*x)` seems to covert to another data structure again
- call the `sum()` function on the set, dead simple; a built-in function that knows how to operate on a set of numbers

The final notation thing looks like typing shenanigans again, and whatever the `.0` does - no idea.

I hope I could provide some ideas. :)

And now for the outcome: If the two implementations give different sums, well... 🙃

Reply:
In other words: If Rust was super duper smart, you could just call `data.sum()`.

Reply:
The whole `Wrapping` bit is to specify that addition should wrap on overflow instead of saturating or (default) panic on overflow.

The `.0` is to extract from `Wrapping(u8)` to `u8`.

An alternative to using a `Wrapping` type is to call `wrapping_add()`. It might be more readable here than this extra wrapping and unwrapping (in the type sense) that we see here.

Reply:
Wrapping is an add that won't overflow. 
the .0 is a notation meaning 'first element in tuple' I believe.

Doesn't matter much, it's more complex than can be done in nostd.